linera-core/src/client/mod.rs

```diff
@@ -713,50 +711,39 @@ where
         if info.next_block_height != BlockHeight::ZERO && self.options.message_policy.is_ignore() {
             return Ok(Vec::new()); // OpenChain is already received, others are ignored.
         }
-        let mut requested_pending_message_bundles = info.requested_pending_message_bundles;
-        let mut pending_message_bundles = vec![];
-        // The first incoming message of any child chain must be `OpenChain`. We must have it in
-        // our inbox, and include it before all other messages.
-        if info.next_block_height == BlockHeight::ZERO
-            && info
-                .description
-                .ok_or_else(|| LocalNodeError::InactiveChain(self.chain_id))?
-                .is_child()
-        {
-            let Some(index) = requested_pending_message_bundles
-                .iter()
-                .position(|message| {
-                    matches!(
-                        message.bundle.messages.first(),
-                        Some(PostedMessage {
-                            message: Message::System(SystemMessage::OpenChain(_)),
-                            ..
-                        })
-                    )
-                })
-            else {
+
+        let needs_openchain_message = info.needs_opening()?;
+        let mut rearranged = false;
+        let mut pending_message_bundles = info.requested_pending_message_bundles;
+
+        if needs_openchain_message {
+            // The first incoming message of any child chain must be `OpenChain`. We must have it in
+            // our inbox, and include it before all other messages.
+            rearranged = IncomingBundle::put_openchain_at_front(&mut pending_message_bundles);
+            if !rearranged {
                 return Err(LocalNodeError::InactiveChain(self.chain_id).into());
             };
```

Comment:
Maybe `ensure!(rearranged, …)` is slightly more concise.
(Especially when we want to ensure a negation it might be easier to read.)

Reply:
Done https://github.com/linera-io/linera-protocol/pull/2745/commits/b5c6ebf561827dffa2aa7d036a59f80549f54b87 .